src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs

```diff
@@ -69,14 +69,51 @@ static async ValueTask<QuicConnection> StartConnectAsync(QuicClientConnectionOpt
 
             using CancellationTokenSource linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
 
-            if (options.HandshakeTimeout != Timeout.InfiniteTimeSpan && options.HandshakeTimeout != TimeSpan.Zero)
-            {
-                linkedCts.CancelAfter(options.HandshakeTimeout);
-            }
-
             try
             {
-                await connection.FinishConnectAsync(options, linkedCts.Token).ConfigureAwait(false);
+                if (!options.RemoteEndPoint.TryParse(out string? host, out IPAddress? address, out int port))
+                {
+                    throw new ArgumentException(SR.Format(SR.net_quic_unsupported_endpoint_type, options.RemoteEndPoint.GetType()), nameof(options));
+                }
+
+                if (address is not null && host is null)
+                {
+                    if (options.HandshakeTimeout != Timeout.InfiniteTimeSpan && options.HandshakeTimeout != TimeSpan.Zero)
+                    {
+                        linkedCts.CancelAfter(options.HandshakeTimeout);
+                    }
```

Comment:
Not sure if this is desired behavior, but currently if we are not successful to connect every ip address that we get from DNS resolution, we're going to throw this:
`System.Net.Sockets.SocketException : A socket operation was attempted to an unreachable host.`

instead of `ConnectionTimeout`

cc @ManickaP

Reply:
IMO unreachable host is better error message than connection timeout. ConnectionTimeout implies we reached the host but the host was slow in completing the connection.

Reply:
I think in sockets the exception depends on the condition. If the port is closed and actively refusing connection via RST or ICMP we would get unreachable error, if it fails on timeout e.g. no response we would throw the timeout. Note that for tests on loopback there is observable difference between Windows and Linux.